# Data Exploration and Wrangling

In [1]:
import pandas as pd
import numpy as np

In [2]:
# import data
collisions = pd.read_csv('accident_data.csv')

C:\Users\mende\AppData\Local\Temp\ipykernel_11356\3500313085.py:2: DtypeWarning: Columns (21,31,38,39,40) have mixed types. Specify dtype option on import or set low_memory=False.
  collisions = pd.read_csv('accident_data.csv')


In [3]:
collisions = collisions.drop('Unnamed: 0', axis = 1) 

In [4]:
collisions.sample(5)

CASE_ID AT_FAULT PARTY_SEX  PARTY_AGE PARTY_SOBRIETY  \
1983884  90939545        N         -        998              H   
2444546  91172746        N         F         53              A   
3692548  91948882        N         M         34              A   
763191    8743839        N         F         27              A   
1061209   8898714        N         M         32              A   

        PARTY_DRUG_PHYSICAL DIR_OF_TRAVEL PARTY_SAFETY_EQUIP_1  \
1983884                   H             E                    -   
2444546                   -             E                    M   
3692548                   -             E                    M   
763191                    -             E                    L   
1061209                   -             W                    M   

        PARTY_SAFETY_EQUIP_2 FINAN_RESPONS  ... LIGHTING CONTROL_DEVICE  \
1983884                    -             O  ...        D              D   
2444546                    G             Y  ...        A              D   
3692548                    G             Y  ...        C              D   
763191                     G             Y  ...        C              A   
1061209                    G             Y  ...        A              A   

        PEDESTRIAN_ACCIDENT BICYCLE_ACCIDENT MOTORCYCLE_ACCIDENT  \
1983884                 NaN              NaN                 NaN   
2444546                 NaN              NaN                 NaN   
3692548                 NaN              NaN                   Y   
763191                  NaN              NaN                 NaN   
1061209                 NaN              NaN                 NaN   

        TRUCK_ACCIDENT  NOT_PRIVATE_PROPERTY ALCOHOL_INVOLVED  LATITUDE  \
1983884            NaN                     Y                Y  34.04851   
2444546            NaN                     Y              NaN  34.06609   
3692548            NaN                     Y              NaN  37.85798   
763191             NaN                     Y              NaN       NaN   
1061209            NaN                     Y              NaN       NaN   

         LONGITUDE  
1983884  117.71044  
2444546  117.20335  
3692548  122.21309  
763191         NaN  
1061209        NaN  

[5 rows x 60 columns]

### Exploratory Analysis

In [5]:
collisions.shape

(4911120, 60)

In [6]:
collisions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4911120 entries, 0 to 4911119
Data columns (total 60 columns):
 #   Column                Dtype  
---  ------                -----  
 0   CASE_ID               int64  
 1   AT_FAULT              object 
 2   PARTY_SEX             object 
 3   PARTY_AGE             int64  
 4   PARTY_SOBRIETY        object 
 5   PARTY_DRUG_PHYSICAL   object 
 6   DIR_OF_TRAVEL         object 
 7   PARTY_SAFETY_EQUIP_1  object 
 8   PARTY_SAFETY_EQUIP_2  object 
 9   FINAN_RESPONS         object 
 10  SP_INFO_2             object 
 11  OAF_VIOLATION_CODE    object 
 12  OAF_VIOL_CAT          object 
 13  OAF_1                 object 
 14  OAF_2                 object 
 15  MOVE_PRE_ACC          object 
 16  VEHICLE_YEAR          float64
 17  VEHICLE_MAKE          object 
 18  STWD_VEHICLE_TYPE     object 
 19  CHP_VEH_TYPE_TOWING   float64
 20  CHP_VEH_TYPE_TOWED    object 
 21  RACE                  object 
 22  COLLISION_DATE        float64
 23  COLLISI

In [7]:
# count na's
na_count = collisions.isna().sum()

na_count[na_count>0]

PARTY_DRUG_PHYSICAL       11159
FINAN_RESPONS            294792
OAF_VIOLATION_CODE       312384
VEHICLE_YEAR             497839
VEHICLE_MAKE                 44
CHP_VEH_TYPE_TOWING      212539
CHP_VEH_TYPE_TOWED      4771117
RACE                     915274
COLLISION_DATE             1825
COLLISION_TIME             1825
REPORTING_DISTRICT      3099574
DAY_OF_WEEK                1825
CHP_SHIFT                  1825
POPULATION                 1825
CNTY_CITY_LOC              1825
BEAT_TYPE                  1825
CHP_BEAT_TYPE              1825
DISTANCE                   1825
DIRECTION               1227949
INTERSECTION               1825
WEATHER_1                  1825
WEATHER_2                  1825
STATE_HWY_IND              7313
LOCATION_TYPE           4779506
RAMP_INTERSECTION       4779506
SIDE_OF_HWY             4779540
PARTY_COUNT                1825
PRIMARY_COLL_FACTOR        1825
PCF_CODE_OF_VIOL           1825
PCF_VIOL_CATEGORY          1825
TYPE_OF_COLLISION          1825
MVIW    

Based on missing values `CHP_VEH_TYPE_TOWED`, `REPORTING_DISTRICT`, `LOCATION_TYPE`,`RAMP_INTERSECTION`, `SIDE_OF_HWY`,
`PEDESTRIAN_ACCIDENT`,`BICYCLE_ACCIDENT`,`MOTORCYCLE_ACCIDENT`,`TRUCK_ACCIDENT` will all be removed since most values are missing. `ALCOHOL_INVOLVED` will stay because in general this is a feature we want to consider in a collisions. A lot of rows have 1825 missing values. Given the size of the data frame, dropping the rows assiciated with 1825 won't significantly impact our final model.

In [8]:
# count dash and blanks
dash_blank_count = ((collisions=='-')| (collisions=='')).sum(axis=0)

print(dash_blank_count[dash_blank_count>0])

PARTY_SEX                785833
PARTY_SOBRIETY           121034
PARTY_DRUG_PHYSICAL     3903904
DIR_OF_TRAVEL            140340
PARTY_SAFETY_EQUIP_1     831528
PARTY_SAFETY_EQUIP_2     881516
SP_INFO_2                645982
OAF_VIOLATION_CODE      4598728
OAF_1                    211676
OAF_2                   4815859
MOVE_PRE_ACC              86713
VEHICLE_MAKE            1182253
STWD_VEHICLE_TYPE        375619
REPORTING_DISTRICT            3
INTERSECTION              34877
WEATHER_1                 18250
WEATHER_2               4792476
RAMP_INTERSECTION         57800
PRIMARY_COLL_FACTOR       16284
PCF_CODE_OF_VIOL        4908887
TYPE_OF_COLLISION         30582
MVIW                      21817
PED_ACTION                 3025
ROAD_SURFACE              28284
ROAD_COND_1               23320
ROAD_COND_2             4879286
LIGHTING                  14247
CONTROL_DEVICE            21974
dtype: int64


Based on dashes and blanks `OAF_VIOLATION_CODE`, `OAF_2`,`WEATHER_2`, `PCF_CODE_OF_VIOL`, `ROAD_COND_2` will be removed due to the high number of missing information. 

In [9]:
# check if any rows are all NA
all_na = collisions.isna().sum(axis=1) == 59 # 59 since all rows have a case id

print('The number of rows with all missing values:', len(all_na[all_na].index.tolist()))

The number of rows with all missing values: 0


## Data Cleaning

In [10]:
# update data to drop columns with excessive missing values
col_to_drop = ['CHP_VEH_TYPE_TOWED', 'REPORTING_DISTRICT', 'LOCATION_TYPE',
               'RAMP_INTERSECTION', 'SIDE_OF_HWY', 'PEDESTRIAN_ACCIDENT',
               'BICYCLE_ACCIDENT','MOTORCYCLE_ACCIDENT','TRUCK_ACCIDENT',
               'OAF_VIOLATION_CODE', 'OAF_2','WEATHER_2', 'PCF_CODE_OF_VIOL', 'ROAD_COND_2']

collisions = collisions.drop(col_to_drop, axis = 1)

### Missing values

In [11]:
# fill values for object dtpyes
obj_dtype = collisions.select_dtypes(include='object').columns
collisions[obj_dtype] = collisions[obj_dtype].fillna('Not Stated')


In [12]:
# drop rows with missing values associated with 1825
collisions = collisions.dropna(subset = ['COLLISION_DATE'])


In [13]:
# drop rows with missing latitude and longitude since this information will be important for the tool later in the project
collisions = collisions.dropna(subset = ['LATITUDE'])

In [14]:
# in order to not introduce bias into the data, missing values for VEHICLE_YEAR will not be imputed with the mean. Using KNN or other advanced imputation strategies would not be appropriate since relevant data to impute this information is not available
collisions = collisions.dropna(subset = ['VEHICLE_YEAR'])

In [15]:
# missing values for CHP_VEH_TYPE_TOWING will be replaced with 99 to reflect an unknown value 
collisions['CHP_VEH_TYPE_TOWING'] = collisions['CHP_VEH_TYPE_TOWING'].fillna(99)

In [16]:
# count na's
na_count = collisions.isna().sum()

na_count[na_count>0]

Series([], dtype: int64)

In [17]:
collisions.shape

(3082221, 46)

### Data Manipulation

In [18]:
# fill dashes and blanks with 'Not Stated' since they mean the same in the data dictionary
def replace_dash_blank(df, replacement_value):
        new_df = df.replace(['-','',' ', '- '], replacement_value)
        return new_df

In [19]:
collisions_no_missing = replace_dash_blank(collisions, 'Not Stated')

In [20]:
# replace Not Stated with N in alchohol_involved
collisions_no_missing['ALCOHOL_INVOLVED'] = collisions_no_missing['ALCOHOL_INVOLVED'].replace('Not Stated', 'N')

In [21]:
# look at unique values to find values that have to be changed to fit contextual information

for col in collisions_no_missing:
    unique_vals = collisions_no_missing[col].unique()
    print(col,':','\n', unique_vals)

CASE_ID : 
 [6698645 8008483 8008487 ... 9656181 9656225 9656269]
AT_FAULT : 
 ['Y' 'N']
PARTY_SEX : 
 ['M' 'Not Stated' 'F' 'X']
PARTY_AGE : 
 [ 31  44  27 998  41  33  47  25  22  39  64  68  36  32  77  21  55  48
  38  67  56  34  19  58  50  35  52  40  18  37  23  60  62  49  74  65
  72  46  29  57  53  28  78  66  70  24  20  63  30  43  26  45  76  59
  61  17  83  69  16  86  79  71  51  54  12  84  75  87  80  81  42  15
  85  73  88  93  13  89  82  10  91   0  90  92  14 119  95 101  98   1
   2 120  96 121   8  94  11   5   9   7   4  99  97 122 104 117 100   3
 112 103   6 123 118 102 105 110 109 107 113 124 111 114 125 106 108]
PARTY_SOBRIETY : 
 ['B' 'A' 'H' 'G' 'D' 'C' 'Not Stated']
PARTY_DRUG_PHYSICAL : 
 ['Not Stated' 'H' 'G' 'E' 'I' 'F']
DIR_OF_TRAVEL : 
 ['N' 'W' 'S' 'E' 'Not Stated']
PARTY_SAFETY_EQUIP_1 : 
 ['M' 'Not Stated' 'L' 'P' 'N' 'B' 'G' 'C' 'R' 'K' 'H' 'U' 'V' 'X' 'A' 'E'
 'Y' 'W' 'S' 'D' 'Q' 'F' 'J']
PARTY_SAFETY_EQUIP_2 : 
 ['G' 'Not Stated' 'B' 'H' 'C

MVIW : 
 ['C' 'E' 'J' 'G' 'D' 'B' 'I' 'A' 'H' 'F' 'Not Stated' '2' '8' '0' '9' '7'
 '3' '4' '6' '1' '5']
PED_ACTION : 
 ['A' 'E' 'B' 'F' 'D' 'C' 'Not Stated' 'G']
ROAD_SURFACE : 
 ['A' 'B' 'Not Stated' 'D' 'C' 'H']
ROAD_COND_1 : 
 ['H' 'D' 'A' 'C' 'B' 'E' 'G' 'F' 'Not Stated']
LIGHTING : 
 ['C' 'A' 'B' 'D' 'E' 'Not Stated' 'G']
CONTROL_DEVICE : 
 ['A' 'D' 'B' 'Not Stated' 'C' 'E' 'G' 'H']
NOT_PRIVATE_PROPERTY : 
 ['Y']
ALCOHOL_INVOLVED : 
 ['Y' 'N']
LATITUDE : 
 [34.57264 34.2176  34.23991 ... 35.61123 38.95526 39.00209]
LONGITUDE : 
 [118.04491 119.1868  119.17244 ... 118.76186 118.64961 118.6373 ]


##### Notes:
`PARTY_SEX` has a value 'X' that must be changed to 'Not Stated'  
`PARTY_AGE` has a 998 value to represent a missing age. 273364 rows have 998
`VEHICLE_MAKE` has different spellings for the same manufacturer  
`COLLISION_DATE` must be separated to year, day, and month  
`COLLISION_TIME` must be seperated to hour and minute  
`CHP_BEAT_TYPE` has mixed data types  
`MVIW` has numbers and letters when only letters should be present based on the data description  
`NOT_PRIVATE_PROPERTY` is all yes so it will be dropped


In [22]:
# change X to Not Stated in PARTY_SEX
collisions_no_missing['PARTY_SEX'] = collisions_no_missing['PARTY_SEX'].replace('X', 'Not Stated')

In [23]:
# split COLLISION_DATE to year, month, day
collisions_no_missing['COLLISION_DATE'] = collisions_no_missing['COLLISION_DATE'].astype('string')

collisions_no_missing['COLLISION_YEAR'] = collisions_no_missing['COLLISION_DATE'].str.slice(0,4)
collisions_no_missing['COLLISION_MONTH'] = collisions_no_missing['COLLISION_DATE'].str.slice(4,6)
collisions_no_missing['COLLISION_DAY'] = collisions_no_missing['COLLISION_DATE'].str.slice(6,8)


In [24]:
# split time to hour and minute
collisions_no_missing['COLLISION_TIME'] = collisions_no_missing['COLLISION_TIME'].astype('string')

collisions_no_missing['COLLISION_HOUR'] = collisions_no_missing['COLLISION_TIME'].str.slice(0,2)
collisions_no_missing['COLLISION_MINUTE'] = collisions_no_missing['COLLISION_TIME'].str.slice(2,4)


In [25]:
# drop columns that are no longer needed
collisions_no_missing = collisions_no_missing.drop(['COLLISION_DATE','NOT_PRIVATE_PROPERTY','COLLISION_TIME'], axis=1)

In [26]:
# replace 0 with '0' in CHP_BEAT_TYPE
collisions_no_missing['CHP_BEAT_TYPE'] = collisions_no_missing['CHP_BEAT_TYPE'].replace(0, '0')

In [27]:
# change numbers in MVIW to letter 0 = A, 1 = B,... per the data description
mviw_dict = {'0': 'A', '1': 'B', '2': 'C', '3': 'D', '4': 'E', '5': 'F', '6': 'G', '7': 'H', '8': 'I', '9': 'J'}

collisions_no_missing['MVIW'] = collisions_no_missing['MVIW'].replace(mviw_dict)

In [28]:
# look at the names of vehicle makes
collisions_no_missing['VEHICLE_MAKE'].unique().tolist()

['NISSAN',
 'AUDI',
 'Not Stated',
 'FORD',
 'SATURN',
 'CHEVROLET',
 'TOYOTA',
 'KIA',
 'OLDSMOBILE',
 'RAM',
 'MAZDA',
 'BMW',
 'HONDA',
 'LEXUS',
 'ACURA',
 'MERCEDES-BENZ',
 ' FORD',
 ' TOYOTA',
 ' CHEVROLET',
 ' DODGE',
 '  HONDA',
 'MERCURY',
 'INFINITI',
 'VOLVO',
 'CADILLAC',
 'ISUZU',
 'GMC',
 ' CHRYSLER',
 'DODGE',
 ' HONDA',
 'SUBARU',
 'MITSUBISHI',
 'HYUNDAI',
 'BUICK',
 'VOLKSWAGEN',
 'CHRYSLER',
 'SUZUKI',
 'JEEP',
 'FIAT',
 'HARLEY DAVIDSON',
 'YAMAHA',
 'JAGUAR',
 'TRIUMPH',
 'SCION',
 'TESLA',
 'PONTIAC',
 'HYPER',
 'FREIGHTLINER',
 'STATE BICYCLE',
 'PETERBUILT',
 'INTERNATIONAL',
 'GROVE',
 'LINCOLN',
 'MINI',
 'KMM KOVATCH',
 'NEWFLYER',
 'PLYMOUTH',
 'KAWASAKI',
 'KENWORTH',
 ' GMC',
 'BIKE',
 'LAND ROVER',
 'SURGE',
 'PORSCHE',
 'CHEV',
 'KIA ',
 'ACUR',
 'VESP',
 'TOYT',
 'DODG',
 'BMW ',
 'HD  ',
 'MAZD',
 'SCIO',
 'MERZ',
 'NISS',
 'LEXS',
 'HOND',
 'SAFARI',
 'VOLK',
 'GMC ',
 'NWFLR',
 'TESL',
 'INFI',
 'HYUN',
 'LNDR',
 'YAMA',
 'SUBA',
 'GIAN',
 'NFLY',
 '

In [29]:
# change incorrect vehicle spelling to correct spelling in vehicle_make column
car_makes = {r'.*for.*': 'Ford', r'.*toy.*': 'Toyota', r'.*sub.*': 'Subaru', r'.*nis.*': 'Nissan', r'.*at.*': 'Saturn',
             r'.*chev.*': 'Chevrolet', r'.*kia.*': 'Kia', r'.*old.*': 'Oldsmobile', r'.*ram.*': 'Dodge', r'.*dod.*': 'Dodge',
             r'.*maz.*': 'Mazda', r'.*hon.*': 'Honda', r'.*lex.*': 'Lexus', r'.*acu.*': 'Acura', r'.*rced.*': 'Mercedes',
             r'.*rcu.*': 'Mercury', r'.*inf.*': 'Infinity', r'.*vol.*': 'Volvo', r'.*cad.*': 'Cadillac', r'.*isu.*': 'Isuzu',
             r'.*chr.*': 'Chrysler',r'.*hyu.*': 'Hyundai',r'.*bui.*': 'Buick',r'.*wag.*': 'Volkswagen',r'.*suz.*': 'Suzuki',
             r'.*gmc.*': 'GMC',r'.*jeep.*': 'Jeep',r'.*fiat.*': 'Fiat',r'.*Har.*': 'Harley Davidson',r'.*yam.*': 'Yamaha',
             r'.*jag.*': 'Jaguar',r'.*tri.*': 'Triumph',r'.*sci.*': 'Scion',r'.*tes.*': 'Tesla',r'.*tsla.*': 'Tesla',
             r'.*pon.*': 'Pontiac',r'.*bic.*': 'Bicycle',r'.*bike.*': 'Bicycle',r'.*frei.*': 'FreightLiner',r'.*peter.*': 'Peterbuilt',
             r'.*int.*': 'International',r'.*lin.*': 'Lincoln',r'.*mini.*': 'Mini',r'.*newf.*': 'Newflyer',r'.*ply.*': 'Plymouth',
             r'.*kaw.*': 'Kawasaki',r'.*ken.*': 'Kenworth',r'.*land.*': 'Land Rover',r'.*por.*': 'Porsche', 'HD': 'Harley Davidson',
             'MERZ': 'Mercedez','VOLK': 'Volkswagen','NWFLR': 'Newflyer','LNDR': 'Land Rover','NEW FLYER': 'Newflyer','SCHWINN': 'Bike',
             'SAAB': 'Saab','TSMR': 'Tesla', r'.*winne.*': 'Winnebago',r'.*hum.*': 'Hummer','STRN': 'Saturn','FRHT': 'FreightLiner',
             r'.*fer.*': 'Ferrari','PTRB': 'Peterbuilt',r'.*range.*': 'Range Rover','NEWFL': 'Newflyer','TRAVERSE':'Chevrolet',
             'thom': 'Thomas Built', 'dat': 'Datsun',r'.*kw.*': 'Kawasaki',r'.*duc.*': 'Ducati','SCHW':'Bike', 'GM': 'GMC', 
            r'.*alf.*': 'Alfa Romeo','NWFRL':'Newflyer', 'auto':'Not Stated', r'.*bm.*': 'BMW', r'.*bent.*': 'Bentley',
             'TYMC': 'Toyota',r'.*grum.*': 'Grumman','TREK': 'Subaru',r'.*mase.*': 'Maserati', r'.*will.*': 'Jeep', 
             'VW': 'Volkswagen','FRGHT':'FreightLiner', 'BB':'Blue Bird'}

collisions_no_missing['VEHICLE_MAKE'] = collisions_no_missing['VEHICLE_MAKE'].astype('string')

for pattern, replacement in car_makes.items():
    collisions_no_missing['VEHICLE_MAKE'] = collisions_no_missing['VEHICLE_MAKE'].str.replace(pattern, replacement, case = False, regex = True)

In [30]:
# check how many vehicle manufacturers there are
print(f"Number of unique manufacturers: \n {len(collisions_no_missing['VEHICLE_MAKE'].unique().tolist())}")

Number of unique manufacturers: 
 4808


There are still too many different spellings for car makes. Replacing with regular expressions will take to long to process this information using thefuzz. After some research, I found a package called Levenshtein that can help group similar strings to a desired string. The following code uses thefuzz to change incorrect car make spelling.

In [31]:
from thefuzz import fuzz, process
# make list of correctly spelled car makes
manuf = ['Ford','Toyota','Subaru','Nissan','Saturn','Chevrolet','Kia','Oldsmobile','Dodge','Mazda','Honda','Lexus',
             'Acura','Mercedes','Mercury','Infinity','Volvo','Cadillac','Isuzu','Chrysler','Hyundai','Buick','Volkswagen',
             'Suzuki','GMC','Jeep','Fiat','Harley Davidson','Yamaha','Jaguar','Triumph','Scion','Tesla','Pontiac','Bicycle',
             'FreightLiner','Peterbuilt','International','Lincoln','Mini Cooper','Newflyer','Plymouth','Mercedez','Land Rover',
             'Saab','Winnebago','Hummer','Ferrari','Range Rover','Thomas Built','Datsun','Kawasaki','Ducati','GMC','Alfa Romeo',
             'BMW','Bentley','Grumman','Maserati','Kenworth','Lucid','Rivian','Nikola','Polaris','Can-Am','John Deere','Smart',
             'Geo','KTM','Caterpillar','Unknown Trailer','Victory','Not Stated','Daewoo','Rolls-Royce','Lamborghini', 'Unknown Motorcycle',
             'Bus','Electric','Genesis','McLaren','Audi', 'Mitsubishi', 'Porsche', 'Blue Bird', 'Gillig','El Dorado','Mack',
             'Fleetwood']

# create function that can be used for any data frame
def replace_misspled_words(df, col, correct_words, scorer, similarity):
    # make copy of df to not change original df
    new_df = df.copy()
    # give a matched word and score for every row in the col
    def correct_word(x):
        matched, score = process.extractOne(x, correct_words, scorer = scorer)
        if score > similarity: 
            return matched
        else:
            return x
    # apply correct_word to new_df
    new_df[col] = new_df[col].apply(correct_word)
    return new_df


In [32]:
# replace words with replace_misspled_words functions
collisions_clean_make = replace_misspled_words(collisions_no_missing, 'VEHICLE_MAKE',manuf,fuzz.WRatio, 60)

In [33]:
# view the number of unique car makes
print(f"Unique Car Makes: \n {len(collisions_clean_make['VEHICLE_MAKE'].unique().tolist())}")

# count the number of car makes that are not in the manf list
print(f"Count of car makes not in manufacturer list: \n {(~collisions_clean_make['VEHICLE_MAKE'].isin(manuf)).sum()}")

Unique Car Makes: 
 2245
Count of car makes not in manufacturer list: 
 9193


In [34]:
# view the top 20 makes not delagated a correct name
not_in_manf =  ~collisions_clean_make['VEHICLE_MAKE'].isin(manuf)

print(collisions_clean_make[not_in_manf]['VEHICLE_MAKE'].value_counts().head(50))

VEHICLE_MAKE
MIFU                 760
AUTC                 437
WSTR                 318
STRG                 217
WEST                 169
WANCO                160
MBZ                  135
KUBOTA               107
SPCNS                106
PREVOST              105
SPECIALIZED          100
VANHOOL               90
VESP                  73
VESPA                 67
SPECIALI              63
OTHER - DOMESTIC      61
JAYCO                 60
OTHER - MOTORHOME     58
UDXX                  57
HUSQ                  57
HYTR                  55
FRTL                  53
FHRT                  52
HUSQVARNA             51
PREV                  50
FRGT                  49
WSTS                  46
HUFFY                 46
NEW HOLL              43
ORION                 43
PIRC                  42
OTHER - MOPED         41
MONGOOSE              40
PTBR                  38
SMEAL                 37
OSHK                  37
VNHL                  36
VAN HOOL              34
ICRP                  34
WRKCC       

The rest of the makes are either unknown or not a vehicle so, they will be replaced with 'Unknown' to differentiate with 'Not Stated'. This is also only 9345 of 3082221 instances which is .3% of the data. 

In [35]:
# replace the rest of vehicle makes with "Unknown"
mask = ~collisions_clean_make['VEHICLE_MAKE'].isin(manuf)

collisions_clean_make.loc[mask, 'VEHICLE_MAKE'] = 'Unknown'

In [36]:
# view final list of vehicle makes
collisions_clean_make['VEHICLE_MAKE'].unique().tolist()

['Nissan',
 'Audi',
 'Saturn',
 'Ford',
 'Chevrolet',
 'Toyota',
 'Kia',
 'Oldsmobile',
 'Dodge',
 'Mazda',
 'BMW',
 'Honda',
 'Lexus',
 'Acura',
 'Mercedes',
 'Mercury',
 'Infinity',
 'Volvo',
 'Cadillac',
 'Suzuki',
 'GMC',
 'Chrysler',
 'Subaru',
 'Hyundai',
 'Buick',
 'Jeep',
 'Harley Davidson',
 'Yamaha',
 'Jaguar',
 'Triumph',
 'Scion',
 'Tesla',
 'Pontiac',
 'Mini Cooper',
 'Lincoln',
 'Land Rover',
 'Newflyer',
 'Plymouth',
 'Kawasaki',
 'Kenworth',
 'Bicycle',
 'Geo',
 'Porsche',
 'Unknown',
 'Mercedez',
 'Saab',
 'Rivian',
 'Unknown Trailer',
 'Can-Am',
 'John Deere',
 'Gillig',
 'Winnebago',
 'Smart',
 'Peterbuilt',
 'Mitsubishi',
 'Hummer',
 'FreightLiner',
 'Ferrari',
 'Thomas Built',
 'Not Stated',
 'Victory',
 'Range Rover',
 'Fleetwood',
 'KTM',
 'Volkswagen',
 'Unknown Motorcycle',
 'International',
 'Alfa Romeo',
 'Grumman',
 'Maserati',
 'Bentley',
 'Electric',
 'Bus',
 'Mack',
 'Genesis',
 'McLaren',
 'Daewoo',
 'Polaris',
 'Rolls-Royce',
 'Blue Bird',
 'Ducati',
 '

In [39]:
# view counts for each age
print(collisions_clean_make['PARTY_AGE'].value_counts())

PARTY_AGE
998    273364
22      79075
24      78653
23      78564
25      78449
        ...  
124         2
111         1
125         1
106         1
108         1
Name: count, Length: 125, dtype: int64


Replacing 998 with the average or median would not be ideal since there are an overwhelming amount of missing ages and that could affect the statistical analysis we will do later. Instead, we will impute these values using Iterative imputer and relavent features such as `VEHICLE_MAKE` and `COLLISION_HOUR`. Since the data set is large, KNN Imputer would take too long to complete the task. However, it's still important to use the least amount of features but highly relavent. If time allows at the end of the project, this is an area we can improve by using correlation to select the most important columns to impute age.


In [41]:
# choose subset of columns for features
columns = ['PARTY_AGE','PARTY_SEX','VEHICLE_MAKE','VEHICLE_YEAR','COLLISION_HOUR','RACE']

age_data = collisions_clean_make[columns]

In [59]:
age_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3082221 entries, 8 to 4911113
Data columns (total 6 columns):
 #   Column          Dtype  
---  ------          -----  
 0   PARTY_AGE       float64
 1   PARTY_SEX       object 
 2   VEHICLE_MAKE    object 
 3   VEHICLE_YEAR    float64
 4   COLLISION_HOUR  float64
 5   RACE            object 
dtypes: float64(3), object(3)
memory usage: 164.6+ MB


In [58]:
# change dtypes 
age_data.loc[:, 'COLLISION_HOUR'] = age_data['COLLISION_HOUR'].astype('float64')

In [46]:
# change 998 to NA
age_data.loc[age_data['PARTY_AGE'] == 998, 'PARTY_AGE'] = np.nan

In [71]:
# one hot encode categorical data
from sklearn.preprocessing import OneHotEncoder

# select columns to one hot encode
cat_cols = ['PARTY_SEX','VEHICLE_MAKE','RACE']

# instantiate object 
ohe = OneHotEncoder(sparse_output = False)

# fit transform the data
encoded_data = ohe.fit_transform(age_data[cat_cols])

# get columns to make a dataframe
columns = ohe.get_feature_names_out()

encoded_df = pd.DataFrame(encoded_data, columns=columns)

# join encoded data frame with original age_data
age_data_drop_cat = age_data.drop(cat_cols, axis = 1)
encoded_age_df = pd.concat([age_data_drop_cat, encoded_df], axis = 1)

In [ ]:
# impute Missing Values with IterativeImputer
from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import IterativeImputer

imputer = IterativeImputer()

imputed_age_data = imputer.fit_transform(encoded_age_df)

imputed_age_df = pd.DataFrame(imputed_age_data, columns = encoded_age_df.columns)


In [ ]:
encoded_age_df

In [40]:
collisions_clean_make.columns

Index(['CASE_ID', 'AT_FAULT', 'PARTY_SEX', 'PARTY_AGE', 'PARTY_SOBRIETY',
       'PARTY_DRUG_PHYSICAL', 'DIR_OF_TRAVEL', 'PARTY_SAFETY_EQUIP_1',
       'PARTY_SAFETY_EQUIP_2', 'FINAN_RESPONS', 'SP_INFO_2', 'OAF_VIOL_CAT',
       'OAF_1', 'MOVE_PRE_ACC', 'VEHICLE_YEAR', 'VEHICLE_MAKE',
       'STWD_VEHICLE_TYPE', 'CHP_VEH_TYPE_TOWING', 'RACE', 'DAY_OF_WEEK',
       'CHP_SHIFT', 'POPULATION', 'CNTY_CITY_LOC', 'BEAT_TYPE',
       'CHP_BEAT_TYPE', 'DISTANCE', 'DIRECTION', 'INTERSECTION', 'WEATHER_1',
       'STATE_HWY_IND', 'PARTY_COUNT', 'PRIMARY_COLL_FACTOR',
       'PCF_VIOL_CATEGORY', 'TYPE_OF_COLLISION', 'MVIW', 'PED_ACTION',
       'ROAD_SURFACE', 'ROAD_COND_1', 'LIGHTING', 'CONTROL_DEVICE',
       'ALCOHOL_INVOLVED', 'LATITUDE', 'LONGITUDE', 'COLLISION_YEAR',
       'COLLISION_MONTH', 'COLLISION_DAY', 'COLLISION_HOUR',
       'COLLISION_MINUTE'],
      dtype='object')